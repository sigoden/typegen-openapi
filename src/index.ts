import * as lodashMerge from "lodash.merge";
import { pascalCase } from "change-case";
import * as lodashGet from "lodash.get";

const METHODS = ["get", "put", "delete", "post", "options"];
const PARAMETER_MAP = { header: "headers", query: "query", path: "params" };


function generate(spec: any, options: Options = {}) {
  const defaultOptions: Options = {
    indent: 2,
  };
  options = lodashMerge(defaultOptions, options);
  const { operations, components } = retriveSchema(spec);
  const builder = new Builder(options);
  for (const operationId in operations) {
    builder.build(pascalCase(operationId + "Req"), operations[operationId]);
  }
  for (const name in components) {
    builder.build(pascalCase(name), components[name]);
  }
  return builder.buffer;
}


class Builder {
  public buffer: string = "";
  private options: Options;
  private indent: number = 0;
  private scopes: string[] = [];
  constructor(options: Options) {
    this.options = options;
  }
  public build(name, schema) {
    if (schema.$ref) {
      this.writeln(`export type ${pascalCase(name)} = ${refTail(schema.$ref)}`)
    } else if (schema.type === "object") {
      this.writeln(`export interface ${pascalCase(name)} {`);
      this.enterScope("object");
      this.buildProperties(schema.properties, schema.required)
      this.exiteScope(true);
      this.writeln("\n");
    }
  }
  private buildProperties(properties: any[], required: string[]) {
    for (const name in properties) {
      const optional = required.find(v => v === name) ? "" : "?";
      const schema = properties[name];
      const type = getType(schema)
      if (isScalar(type)) {
        this.writeln(`${name}${optional}: ${type};`)
      } else {
        if (type === "array") {
          const elemSchema = schema.items;
          const type = getType(elemSchema);
          if (isScalar(type)) {
            this.writeln(`${name}${optional}: ${type}[];`)
          } else {
            this.writeln(`${name}${optional}: {`)
            this.enterScope(type);
            this.buildProperties(elemSchema.properties, elemSchema.required)
            this.exiteScope();
          }
        } else {
          this.writeln(`${name}${optional}: {`)
          this.enterScope(type);
          this.buildProperties(schema.properties, schema.required)
          this.exiteScope();
        }
      }
    }
  }
  private enterScope(kind: string) {
    this.scopes.push(kind);
    this.indent += 1;
  }
  private exiteScope(root = false) {
    const kind = this.scopes.pop();
    const semi = root ? "" : ";";
    this.indent -= 1;
    if (kind === "object") {
      this.writeln(`}${semi}`);
    } else {
      this.writeln(`}[]${semi}`)
    }
  }
  private writeln(line) {
    this.buffer += this.spaces() + line + "\n";
  }
  private spaces() {
    return " ".repeat(this.indent * this.options.indent)
  }
}


function retriveSchema(spec: any) {
  const schemas = {} as any;
  for (const path in spec.paths) {
    const pathItem = spec.paths[path];
    for (const method of METHODS) {
      const operation = pathItem[method]; 
      if (!operation) continue;
      if (!operation.operationId) {
        throw new Error(`endpoint ${method.toUpperCase()} ${path} miss operationId`);
      }
      const endpointSchema = createDefaultSchema();
      const addParamaterSchema = (key, obj: any) => {
        const dataKey = PARAMETER_MAP[key];
        if (!dataKey) return;
        let data = endpointSchema.properties[dataKey];
        if (!data) {
          data = endpointSchema.properties[dataKey] = createDefaultSchema();
        }
        if (obj.$ref) {
          const paths = refToPath(obj.$ref);
          if (paths) {
            data.properties[obj.name] = lodashGet(spec, paths.concat("schema"));
          } else {
            data.properties[obj.name] = {};
          }
        } else {
          data.properties[obj.name] = obj.schema;
        }
        if (obj.required) data.required.push(obj.name);
      };
      const parameters: any[] = [...(pathItem.parameters || []), ...(operation.parameters || [])];
      for (const parameter of parameters) {
        addParamaterSchema(parameter.in, parameter);
      }
      const bodySchema = lodashGet(operation, ["requestBody", "content", "application/json", "schema"]);
      if (bodySchema) {
        endpointSchema.properties["body"] = bodySchema;
      }
      schemas[operation.operationId] = endpointSchema;
    }
  }
  return { operations: schemas, components: lodashGet(spec, "components.schemas") };
}


function createDefaultSchema() {
  return { type: "object", properties: {}, required: [] };
}

function refToPath(ref: string) {
  if (ref === "#") return;
  if (!ref.startsWith("#/")) return;
  return ref.slice(2).split("/");
}

function refTail(ref: string) {
  const paths = refToPath(ref);
  if (paths) return paths[paths.length - 1];
  return "any";
}

function isScalar(type: string) {
  return ["object", "array"].indexOf(type) === -1;
}

function getType(shcema: any) {
  switch (shcema.type) {
    case "integer":
      return "number";
    case "number":
    case "string":
    case "boolean":
    case "object":
    case "array":
      return shcema.type
    default:
      if (shcema.properties) {
        return "object";
      } else if (shcema.items) {
        return "array"
      } else if (shcema.$ref) {
        return refTail(shcema.$ref);
      }
      return "any";
  }
}

export interface Options {
  /**
   * Ident with spaces, default value is 2
   */
  indent?: number;
}

export { generate };
