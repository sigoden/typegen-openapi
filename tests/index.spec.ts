import { generate } from "../src";


test("should works", () => {
  const petstore = require("./spec/cases.json"); // eslint-disable-line
  const result = generate(petstore, {});
  expect(result).toMatchSnapshot();
});
