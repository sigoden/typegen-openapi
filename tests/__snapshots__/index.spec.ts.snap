// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should works 1`] = `
"export interface UpdatePetReq {
  body: Pet;
}

export interface AddPetReq {
  body: Pet;
}

export interface FindPetsByStatusReq {
  query: {
    status?: string;
  };
}

export interface FindPetsByTagsReq {
  query: {
    status?: string[];
  };
}

export interface GetPetByIdReq {
  params: {
    petId: number;
  };
}

export interface DeletePetReq {
  params: {
    petId: number;
  };
  headers: {
    api_key: string;
  };
}

export interface UpdatePetWithFormReq {
  params: {
    petId: number;
  };
  query: {
    name: string;
    status: string;
  };
}

export interface UploadPetImageReq {
  params: {
    petId: number;
  };
  query: {
    additionalMetadata: string;
  };
}

export interface GetInventoryReq {
}

export interface PlaceOrderReq {
  body: Order;
}

export interface GetOrderByIdReq {
  params: {
    orderId: number;
  };
}

export interface DeleteOrderReq {
  params: {
    orderId: number;
  };
}

export interface CreateUserReq {
  body: User;
}

export interface CreateUsersWithListInputReq {
  body: User[];
}

export interface LoginUserReq {
  query: {
    username: string;
    password: string;
  };
}

export interface LogoutUserReq {
}

export interface GetUserByNameReq {
  params: {
    username: string;
  };
}

export interface UpdateUserReq {
  params: {
    username: string;
  };
  body: {
  };
}

export interface DeleteUserReq {
  params: {
    username: string;
  };
}

export interface Category {
  id: number;
  name: string;
}

export interface Pet {
  id: number;
  name: string;
  category: Category;
  photoUrls: string[];
  tags: Tag;
  status: string;
}

export interface ApiResponse {
  code: number;
  message: string;
}

export interface Order {
  id: number;
  petId: number;
  quantity: number;
  shipDate: string;
  status: string;
  complete: boolean;
}

export interface User {
  id: number;
  username: string;
  firstName: string;
  lastName: string;
  email: string;
  password: string;
  phone: string;
  userStatus: number;
}
"
`;
